src/ripple/app/misc/AmendmentTableImpl.cpp

```diff
@@ -97,51 +101,110 @@ class AmendmentTableImpl : public AmendmentTable
     amendmentList_t getDesired();    // amendments we support, do not veto, are not enabled
 };
 
-void
-AmendmentTableImpl::addInitial ()
+namespace detail
 {
-    // For each amendment this version supports, construct the AmendmentState object by calling
-    // addKnown. Set any vetoes or defaults. A pointer to the AmendmentState can be stashed
+using PreEnabledAmendmentsCollection = std::vector<AmendmentName>;
+
+/** preEnabledAmendments is a static collection of amendments are are enabled at build time.
+
+   Add amendments to this collection at build time to enable them on this server.
+*/
+
+PreEnabledAmendmentsCollection const preEnabledAmendments;
 }
 
-AmendmentState*
-AmendmentTableImpl::getCreate (uint256 const& amendmentHash, bool create)
+void
+AmendmentTableImpl::addInitial (Section const& section)
 {
-    // call with the mutex held
-    auto iter (m_amendmentMap.find (amendmentHash));
-
-    if (iter == m_amendmentMap.end())
+    for (auto const& a : detail::preEnabledAmendments)
     {
-        if (!create)
-            return nullptr;
+        if (!a.valid ())
+        {
+            std::string const errorMsg =
+                (boost::format (
+                     "preEnabledAmendments contains an invalid hash (expected "
+                     "a hex number). Value was: %1%") %
+                 a.hexString ()).str ();
+            throw std::runtime_error (errorMsg);
+        }
+    }
 
-        AmendmentState* amendment = & (m_amendmentMap[amendmentHash]);
+    detail::PreEnabledAmendmentsCollection toAdd (detail::preEnabledAmendments);
 
+    {
+        // add the amendments from the config file
+        int const numExpectedToks = 2;
+        for (auto const& line : section.lines ())
         {
-            std::string query = "SELECT FirstMajority,LastMajority FROM Features WHERE hash='";
-            query.append (to_string (amendmentHash));
-            query.append ("';");
-
-            auto sl (getApp().getWalletDB ().lock ());
-            auto db = getApp().getWalletDB ().getDB ();
+            boost::tokenizer<> tokenizer (line);
+            std::vector<std::string> tokens (tokenizer.begin (),
+                                             tokenizer.end ());
+            if (tokens.size () != numExpectedToks)
+            {
+                std::string const errorMsg =
+                    (boost::format (
+                         "The %1% section in the config file expects %2% "
+                         "items. Found %3%. Line was: %4%") %
+                     SECTION_AMENDMENTS % numExpectedToks % tokens.size () %
+                     line).str ();
+                throw std::runtime_error (errorMsg);
+            }
 
-            if (db->executeSQL (query) && db->startIterRows ())
+            toAdd.emplace_back (std::move (tokens[0]), std::move (tokens[1]));
```

Comment:
Nice use of `emplace`.  I need to develop that habit.

I also find this code much easier to read than the original nested `for`.  Good change.